app/service/SubscriptionFieldsService.scala

```diff
@@ -38,12 +47,68 @@ class SubscriptionFieldsService @Inject()(connectorsWrapper: ConnectorsWrapper)(
     }
   }
 
-  def saveFieldValues(application: Application, apiContext: String, apiVersion: String, fields: Fields)
-                     (implicit hc: HeaderCarrier): Future[SaveSubscriptionFieldsResponse] = {
+  def saveFieldValues( role : Role,
+                        application : Application,
+                        apiContext: String,
+                        apiVersion : String,
+                        oldValues: Seq[SubscriptionFieldValue],
+                        newValues: Map[String, String])
+                        (implicit hc: HeaderCarrier) : Future[ServiceSaveSubscriptionFieldsResponse] = {
+    case class AccessDenied()
+    if (newValues.isEmpty) {
+        Future.successful(SaveSubscriptionFieldsSuccessResponse)
+    } else {
 
-    val connector = connectorsWrapper.forEnvironment(application.deployedTo).apiSubscriptionFieldsConnector
+      def toNewValue(oldValue: SubscriptionFieldValue)(newFormValue: String) = {
+        if (oldValue.definition.access.devhub.satisfiesWrite(DevhubAccessLevel.fromRole(role))){
+          Right(oldValue.copy(value = newFormValue))
+        } else {
+          Left(AccessDenied())
+        }
+      }
+
+      val eitherValuesToSave = oldValues.map(oldValue => 
+        newValues.get(oldValue.definition.name) match {
+          case Some(newFormValue) => toNewValue(oldValue)(newFormValue)
+          case None => Right(oldValue)
+        }
+      )
+
+      def doConnectorSave(valuesToSave: Seq[SubscriptionFieldValue]) = { 
+        val connector = connectorsWrapper.forEnvironment(application.deployedTo).apiSubscriptionFieldsConnector
+        val fieldsToSave = valuesToSave.map(v => (v.definition.name -> v.value)).toMap
+
+        connector.saveFieldValues(application.clientId, apiContext, apiVersion, fieldsToSave)
+      }
+
+      eitherValuesToSave.toList.sequence.fold(
+        accessDenied => Future.successful(SaveSubscriptionFieldsAccessDeniedResponse),
```

Comment:
accessDenied appears to be unused

Reply:
We did consider using an Option instead of an Either as nothing cares why it 'fails'. But we thought it was more descriptive of why it failed (AccessDenied). We did wonder if the AccessRequirement should expose the 'allowed-ness' as a Monad - but realised we didn't really know how!

What's your preference: Use Option or leave an Either?

Reply:
The more I look at saveFieldValues, the more I wanna re-write the whole thing :(

Reply:
Discussed with Pete on a call. Small tweaks made